stage1_fly/process-credentials.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stage1_fly
+
+import (
+	"fmt"
+	"runtime"
+	"syscall"
+
+	"github.com/appc/spec/schema"
+	"github.com/hashicorp/errwrap"
+	rktlog "github.com/rkt/rkt/pkg/log"
+	"github.com/rkt/rkt/pkg/user"
+)
+
+type ProcessCredentials struct {
+	uid               int
+	gid               int
+	supplementaryGIDs []int
+}
+
+func LookupProcessCredentials(ra *schema.RuntimeApp, rfs string) (*ProcessCredentials, error) {
+	var c ProcessCredentials
+
+	uidResolver, err := user.NumericIDs(ra.App.User)
+	if err != nil {
+		uidResolver, err = user.IDsFromStat(rfs, ra.App.User, nil)
+	}
+
+	if err != nil { // give up
+		return nil, errwrap.Wrap(fmt.Errorf("invalid user %q", ra.App.User), err)
+	}
+
+	if c.uid, _, err = uidResolver.IDs(); err != nil {
+		return nil, errwrap.Wrap(fmt.Errorf("failed to configure user %q", ra.App.User), err)
+	}
+
+	gidResolver, err := user.NumericIDs(ra.App.Group)
+	if err != nil {
+		gidResolver, err = user.IDsFromStat(rfs, ra.App.Group, nil)
+	}
+
+	if err != nil { // give up
+		return nil, errwrap.Wrap(fmt.Errorf("invalid group %q", ra.App.Group), err)
+	}
+
+	if _, c.gid, err = gidResolver.IDs(); err != nil {
+		return nil, errwrap.Wrap(fmt.Errorf("failed to configure group %q", ra.App.Group), err)
+	}
+
+	// supplementary groups - ensure primary group is included
+	n := len(ra.App.SupplementaryGIDs)
+	c.supplementaryGIDs = make([]int, n+1, n+1)
+	c.supplementaryGIDs[0] = c.gid
+	copy(c.supplementaryGIDs[1:], ra.App.SupplementaryGIDs)
+
+	return &c, nil
+}
+
+func SetProcessCredentials(c *ProcessCredentials, diag *rktlog.Logger) error {
+	// lock the current goroutine to its current OS thread.
+	// This will force the subsequent syscalls to be executed in the same OS thread as Setresuid, and Setresgid,
+	// see https://github.com/golang/go/issues/1435#issuecomment-66054163.
+	runtime.LockOSThread()
```

Comment:
I think this should be moved as one of the first steps in "enter" and "run" and here just mention that it affects only the current thread and may misbehave with free goroutines.